babel/units.py

```diff
@@ -0,0 +1,241 @@
+    """
+    Expand an unit into a qualified form.
+
+    Known units can be found in the CLDR Unit Validity XML file:
+    http://unicode.org/repos/cldr/tags/latest/common/validity/unit.xml
+
+    >>> _find_unit_pattern("radian", locale="en")
+    'angle-radian'
+
+    Unknown values will return None.
+
+    >>> _find_unit_pattern("horse", locale="en")
+
+    :param unit_id: the code of a measurement unit.
+    :return: A key to the `unit_patterns` mapping, or None.
+    """
+    locale = Locale.parse(locale)
+    unit_patterns = locale._data["unit_patterns"]
+    if unit_id in unit_patterns:
+        return unit_id
+    for unit_pattern in sorted(unit_patterns, key=len):
+        if unit_pattern.endswith(unit_id):
+            return unit_pattern
+
+
+def format_unit(value, measurement_unit, length='long', format=None, locale=LC_NUMERIC):
+    """Format a value of a given unit.
+
+    Values are formatted according to the locale's usual pluralization rules
+    and number formats.
+
+    >>> format_unit(12, 'length-meter', locale='ro_RO')
+    u'12 metri'
+    >>> format_unit(15.5, 'length-mile', locale='fi_FI')
+    u'15,5 mailia'
+    >>> format_unit(1200, 'pressure-inch-hg', locale='nb')
+    u'1\\xa0200 tommer kvikks\\xf8lv'
+
+    Number formats may be overridden with the ``format`` parameter.
+
+    >>> from babel._compat import Decimal
+    >>> format_unit(Decimal("-42.774"), 'temperature-celsius', 'short', format='#.0', locale='fr')
+    u'-42,8 \\xb0C'
+
+    The locale's usual pluralization rules are respected.
+
+    >>> format_unit(1, 'length-meter', locale='ro_RO')
+    u'1 metru'
+    >>> format_unit(0, 'length-picometer', locale='cy')
+    u'0 picometr'
+    >>> format_unit(2, 'length-picometer', locale='cy')
+    u'2 bicometr'
+    >>> format_unit(3, 'length-picometer', locale='cy')
+    u'3 phicometr'
+
+    >>> format_unit(15, 'length-horse', locale='fi')
+    Traceback (most recent call last):
+        ...
+    UnknownUnitError: length-horse is not a known unit in fi
+
+    .. versionadded:: 2.2.0
+
+    :param value: the value to format.
+    :param measurement_unit: the code of a measurement unit.
+                             Known units can be found in the CLDR Unit Validity XML file:
+                             http://unicode.org/repos/cldr/tags/latest/common/validity/unit.xml
+    :param length: "short", "long" or "narrow"
+    :param format: An optional format, as accepted by `format_decimal`.
+    :param locale: the `Locale` object or locale identifier
+    """
+    locale = Locale.parse(locale)
+
+    q_unit = _find_unit_pattern(measurement_unit, locale=locale)
+    if not q_unit:
+        raise UnknownUnitError(unit=measurement_unit, locale=locale)
+    unit_patterns = locale._data["unit_patterns"][q_unit].get(length, {})
+    formatted_value = format_decimal(value, format, locale)
+    plural_form = locale.plural_form(value)
+    if plural_form in unit_patterns:
```

Comment:
CLDR here casually mentions length fallbacks (at least that's how I interpreted it here). It'd be awesome to see it here, I don't think it would be that difficult to add an extra iteration through a hardcoded, ordered list of length strings. If it's gross, feel free to punt.